provisioning_templates/snippet/set_hostname.erb

```diff
@@ -0,0 +1,20 @@
+<%#
+name: set_hostname
+snippet: true
+model: ProvisioningTemplate
+kind: snippet
+-%>
+if [ `command -v hostnamectl` ]; then
+  hostnamectl set-hostname <%= @hostname %>
+else
+  hostname <%= @hostname %>
+  # persist the change to survive  reboot
+  echo "<%= @hostname -%>" > /etc/hostname
+fi
+
+if grep '^127.0.1.1\s*<%= @hostname -%>\s*<%= @shortname -%>$' /etc/hosts; then
+  echo "skipping /etc/hosts update, record is already present"
+else
+  echo "adding record to the first line of /etc/hosts"
+  sed -i '1s/^/127.0.1.1   <%= @hostname -%> <%= @shortname -%>\n/' /etc/hosts
```

Comment:
Relying on the OS being correct is what got us into this situation in the first place. If we trusted the contents of `/etc/hosts` we wouldn't need to mess with this at all. Given that we're altering a snippet that may be in use (and in the user_data templates, where hosts could have all sorts of wonky base images), shouldn't we try to preserve behaviour?